lib/internal/modules/esm/resolve.js

```diff
@@ -587,12 +587,35 @@ function shouldBeTreatedAsRelativeOrAbsolutePath(specifier) {
   return false;
 }
 
+function resolveRelativePackage(specifier /* string */, base /* URL */) {
+  const resolved = new URL(specifier, base);
+  const stat = tryStatSync(resolved);
+  if (stat.isDirectory()) {
+    let packageJSONPath;
+    if (!StringPrototypeEndsWith(resolved.pathname, '/')) {
+      packageJSONPath = `${resolved.pathname}/package.json`;
+    } else {
+      packageJSONPath = `${resolved.pathname}package.json`;
+    }
+    const packageJSONUrl = new URL(packageJSONPath, base);
+    packageJSONPath = fileURLToPath(packageJSONUrl);
+    const packageConfig = getPackageConfig(packageJSONPath, base);
+    return packageMainResolve(packageJSONUrl, packageConfig, base);
+  }
+  return resolved;
+}
```

Comment:
I think this should probably combined into the `finalizeResolution` function above rather to maintain full equivalence with the LOAD_AS_FILE | LOAD_AS_DIRECTORY checks per the module resolution spec.

Reply:
The first version I wrote puts this part in `finalizeResolution` function indeed. But I think `finalizeResolution` is `LOAD_AS_FILE` routine, it is called by many other resolve functions as final return call and shouldn't add any `resolve` functionality. So I move this code pieces after the relative path check to fit this rule:
> 3. If X begins with './' or '/' or '../'
>   a. LOAD_AS_FILE(Y + X)
>   b. LOAD_AS_DIRECTORY(Y + X)`

Reply:
It is ok for `finalizeResolution` to include the `LOAD_AS_FILE | LOAD_AS_DIRECTORY` path - the reason for this is that all resolutions eventually end in this check in CommonJS.

Reply:
I don't think it's a good idea...
`finalizeResolution` is pretty simple now, and keep it simple is better to my mind. Merge package resolve logic to it makes it a bit dirty, the call path does not straight anymore.
But if more people think merge logic is better, I will change it...

Reply:
@lingsamuel let me restate my feedback then. This PR only fixes one case, when there are two other test cases not being covered by this PR that you should make pass as well as part of this fix:

1. `import 'pkg/x'` should load `/path/to/node_modules/pkg/x/pjson-main.js` with package.json main checks
2. `import 'pkg/x'` where `pkg/package.json` contains `{ "exports": { "./x": "./z" }` should load `/path/to/node_modules/pkg/z/pjson-main.js`